### PR TITLE
Do not list Android or iOS devices when feature disabled

### LINKED ISSUE #84412: Still see iOS and Android devices after turning off iOS and Android development via "flutter config  --no-enable-ios" and "--no-enable-android"
```
$ flutter config  --no-enable-ios 
Setting "enable-ios" value to "false".

You may need to restart any open editors for them to read new settings.
flutter$ flutter devices                 
1 connected device:

iPad (mobile) • 00008027-001925360106802E • ios • iOS 15.0
```

Since ["ephemeral" devices are preferred](https://github.com/flutter/flutter/pull/34802), this made desktop macOS development really annoying until I unplugged my iPhone.

iOS and Android workflow `canListDevices` do not check if the platform feature flag is enabled:
https://github.com/flutter/flutter/blob/515003e4af54b124054107c803d5ec657196f3f7/packages/flutter_tools/lib/src/ios/ios_workflow.dart#L24-L28
https://github.com/flutter/flutter/blob/f9cd87e60a9bce9343b4e887d0353fe608db29d5/packages/flutter_tools/lib/src/android/android_workflow.dart#L55-L61

Newer platforms do:
https://github.com/flutter/flutter/blob/e7b7ebc066c1b2a5aa5c19f8961307427e0142a6/packages/flutter_tools/lib/src/web/workflow.dart#L29
https://github.com/flutter/flutter/blob/e7b7ebc066c1b2a5aa5c19f8961307427e0142a6/packages/flutter_tools/lib/src/macos/macos_workflow.dart#L27
https://github.com/flutter/flutter/blob/e7b7ebc066c1b2a5aa5c19f8961307427e0142a6/packages/flutter_tools/lib/src/linux/linux_workflow.dart#L30

It seems like `canListDevices ` should only be true if `appliesToHostPlatform` is true.

### PR DESCRIPTION
**Pull request by jmagman**
Respect feature and host platform when deciding whether to list Android and iOS devices.
Fixes https://github.com/flutter/flutter/issues/84412

### CODE REVIEW

File: packages/flutter_tools/test/general.shard/ios/ios_workflow_test.dart [hunk 1]
```diff
@@ -41,16 +45,25 @@ void main() {
     );
 
     expect(iosWorkflow.appliesToHostPlatform, false);
+    expect(iosWorkflow.canLaunchDevices, false);
+    expect(iosWorkflow.canListDevices, false);
   });
 
-  testWithoutContext('iOS workflow is enabled on macOS', () {
+  testWithoutContext('iOS workflow applies on macOS, no Xcode', () {
```
Comment (jmagman) [inline packages/flutter_tools/test/general.shard/ios/ios_workflow_test.dart:52]: The next test was the same set up. Instead, make this one check what happens when Xcode is not installed.

File: packages/flutter_tools/test/general.shard/android/android_workflow_test.dart [hunk 2]
```diff
@@ -73,6 +73,55 @@ void main() {
...
+      operatingSystemUtils: FakeOperatingSystemUtils(),
+    );
+
+    expect(androidWorkflow.appliesToHostPlatform, false);
+    expect(androidWorkflow.canLaunchDevices, false);
+    expect(androidWorkflow.canListDevices, false);
+    expect(androidWorkflow.canListEmulators, false);
+  });
+
+  testWithoutContext('AndroidWorkflow cannot list emulators', () {
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/general.shard/android/android_workflow_test.dart:96]: `... if emulatorPath is null`

**Review by jonahwilliams — APPROVED** [review]
LGTM, thank you for fixing this!

File: packages/flutter_tools/test/general.shard/android/android_workflow_test.dart (hunk 2 above)
Comment (jmagman) [reply packages/flutter_tools/test/general.shard/android/android_workflow_test.dart]: Done.

**Pull request merged**